README
NAME
    README - General Information about WordNet::SenseRelate::AllWords

OVERVIEW
    This module carries out word sense disambiguation (WSD), which is the
    process of selcting the correct sense for a word in a given context. The
    correct sense is selected from a sense inventory which lists the
    possible meanings of a word. This module uses the WordNet lexical
    database as it's sense inventory.

SYNOPSIS
        use WordNet::SenseRelate::AllWords;
        use WordNet::QueryData;
        use WordNet::Tools;

        my $qd = WordNet::QueryData->new;
        my $wntools = WordNet::Tools->new($qd);    
        my %options = (wordnet => $qd,
                       wntools => $wntools,
                       measure => 'WordNet::Similarity::lesk'
                       );

       my $obj = WordNet::SenseRelate::AllWords->new(%options);
       my @context = qw/when in the course of human events/;
       my @res = $obj->disambiguate (window => 3,
                                      scheme => 'normal',
                                      tagged => 0,
                                      context => [@context]);                                    
       print join (' ', @res), "\n";

CONTENTS
    When the distribution is unpacked, several subdirectories are created:

    /lib
        This directory contains the Perl modules that do the actual work of
        disambiguation. By default, these files are installed into
        /usr/local/lib/perl5/site_perl/PERL_VERSION (where PERL_VERSION is
        the version of Perl you are using). See the INSTALL file for more
        information.

    /utils
        This directoy contains a number of scripts that let you run word
        sense disambiguation experiments and reformat data.

        These scripts will be install when 'make install' is run. By
        default, these files are installed into your /usr/local/bin
        directory. See the INSTALL file for more information. The scripts in
        this directory are:

        wsd.pl
            This very useful script can be used to disambiguate a file of
            words. It is discussed in greater detail later in this document.

        semcor-reformat.pl
            This script will reformat a Semcor file so that it can be used
            as input to wsd.pl

        scorer2-format.pl
            This script will reformat the output of wsd.pl so that it can be
            used as input to the Senseval scorer2 program.

        scorer2-sort.pl
            This script will sort the output of scorer2-format.pl so that it
            can be used as input to the Senseval scorer2 program.

        wsd-experiments.pl
            This script will call the above scripts and will run wsd
            experiments.

        allwords-scorer2.pl
            This script is modeled after the senseval scorer2 C program
            (http://www.senseval.org/senseval3/scoring) and will be used for
            scoring.

        sentence_split.pl
            This script splits text into sentences. We expect that the input
            format to allwords should be one sentence per line, one line per
            sentence. If your text is not in this format, you can use this
            script to split the text into sentences. Note that this script
            is not included anywhere and if your text is not in the required
            format, you should call this script explicitly before using
            wsd.pl or the web interface.

        Each of these scripts has detailed documentation. Run perldoc on a
        file to see the detailed documentation; for example, 'perldoc
        wsd.pl' shows the documentation for wsd.pl.

    /doc
        This directory contains all of the *pod files used to document the
        system. These are processed via pod2text and the output of this is
        placed in the top level directory, although these top level text
        files should be considered read only.

    /samples
        This directory contains examples of the different formats of data
        that are supported by this package. It also contains a sample
        stoplist. There is a README file in the directory that describes the
        contents in more detail.

    /t  This directory contains test scripts. These scripts are run when you
        execute 'make test'.

    /web
        This directory contains the allwords web server and interface. There
        are detailed README and INSTALL instructions within this directory.
        Installing the web interface is optional, and is separate from
        installing the main package.

DESCRIPTION
    Words can have multiple meanings or senses. For example, the word
    *glass* in WordNet [1] has seven senses as a noun and five senses as a
    verb. Glass can mean a clear solid, a container for drinking, the
    quantity a drinking container will hold, etc. WSD is the process of
    selecting the correct sense of a word when that word occurs in a
    specific context. For example, in the sentence, "the window is made of
    glass", the correct sense of glass is the first sense, a clear solid.

    WordNet::SenseRelate::AllWords extends a word sense disambiguation
    algorithm described by Pedersen, Banerjee, and Patwardhan [2] by making
    it disambiguate all words in text. The previous version of the algorithm
    was intended for lexical sample data, which means that a single word in
    a context is designated as the target word and is the only word to be
    disambiguated. By contrast, WordNet::SenseRelate::AllWords will assign a
    sense to every word known to WordNet that appears in a context.

    Prior to execution of the algorithm, we remove any word that is not
    known to WordNet, and any word that appears in a stoplist. The input to
    the algorithm is presumed to be a single sentence where non-WordNet
    words and stoplisted words have been removed.
    WordNet::SenseRelate::AllWords does not cross sentence boundaries when
    carrying out disambiguation.

  Algorithm
      for each word w in sentence
        disambiguate-single-word (w)

      disambiguate-single-word (w)
        for each sense s_ti of target word t, where i=0..N
            let score_i = 0

            for each word w_j in context_window 
                next if j = t

                for each sense s_jk of w_j
                    temp-score_k = relatedness (s_ti, s_jk)
                best-score = max temp-score
                if best-score > pairScore
                    score_i = score_i + best-score

        return s_ti s.t. score_i > score_j for all j in 
                    {s_t0, ..., s_tN} and score_i > contextScore

  The Context Window
    The size of the context window can be specified by the user. A context
    window of size 3 means that the context window will consist of three
    words, including the target word. Thus, the three words would be the
    word to the left of the target word, the target word itself, and the
    word to the right of the target word. The algorithm will expand the
    context window so that the three words will be words known to WordNet
    (the algorithm is unable to disambiguate words unknown to WordNet). For
    example, if the word 'the', occurs in the context window to the left of
    the target word, then the window will be expanded by one word to the
    left.

    If the window size is an even number, then there will be one more word
    to the left of the target word than to the right. For example, if the
    window size is 4, there will be two words to the left of the target word
    and one word to the right.

    Note that the context window will only include words in the same
    sentence as the target word. If, for example, the target word is the
    first word in the sentence, then there will be no words to left of the
    target word in the context window regarless of the specified window
    size.

    The minimum window size is 2 because a smaller window mean that there
    are no context words in the window. When the window size is 2, there is
    no context to use for disambiguating the first word in a sentence. To
    assign a sense number to that first word, the first sense of the word is
    chosen (i.e., sense number 1). Sense number 1 is usually the most
    frequent sense of a word.

  Part of Speech Coercion
    Certain measures of semantic similarity only work on noun-noun or
    verb-verb pairs; therefore, the usefulness of these measures for WSD is
    somewhat limited. As a way of coping with this problem,
    WordNet::SenseRelate::AllWords provides an option to "coerce" words in
    the context window to be of the same part of speech as the target word.

    When POS coercion is in effect, if the target word is a noun, then
    WordNet::SenseRelate::AllWords will attempt to convert non-nouns in the
    context window to noun forms of the same word. For example, if the
    target word is a noun and the verb *love* occurs in the window, the
    module might convert that word to the noun *love*.

    WordNet::SenseRelate::AllWords first uses the validForms method from
    WordNet::QueryData to find any valid forms of the word being coerced
    that are of the desired part of speech. In the case of part of speech
    tagged text, the POS tags are discarded. If validForms did not return
    any forms of the desired part of speech, then the derived forms relation
    in WordNet is used to find possible forms of the word. If neither of
    these methods returned usable forms, then no further attempt is made to
    coerce the word to be the desired part of speech.

  Tracing/Debugging
    Several different levels of trace output are available. The trace level
    can be specified as a command-line option to wsd.pl or as a parameter to
    the WordNet::SenseRelate::AllWords module.

   Trace Levels
    The trace levels are:

      1 Show the context window for each pass through the algorithm.

      2 Display winning score for each pass (i.e., for each target word).

      4 Display the non-zero scores for each sense of each target
        word (overrides 2).

      8 Display the non-zero values from the semantic relatedness measures.

     16 Show the zero values as well when combined with either 4 or 8.
        When not used with 4 or 8, this has no effect.

     32 Display traces from the semantic relatedness module.

    Different trace levels can be combined to achieve the desired behavior.
    For example, by specifying a trace level of 3, both level 1 and level 2
    traces are generated (i.e., the context window will be shown along with
    the winning score for each pass).

  Using wsd.pl
    The wsd.pl script provides an easy method of performing disambiguation
    from the command line. The text to be disambiguated is read from a file
    provided by the user on the command line.

   Output
    The output of wsd.pl is simply the disambiguated words. The output will
    be in the form word#part_of_speech#sense_number. The part of speech will
    be one of 'n' for noun, 'v' for verb, 'a' for adjective, or 'r' for
    adverb. Words from other parts of speech are not disambiguated and are
    not found in WordNet. The sense number will be a WordNet sense number.
    WordNet sense numbers are assigned by frequency, so sense 1 of a word is
    more common than sense 2, etc.

    Sometimes when a word is disambiguated, a "different" but synonymous
    word will be found in the output. This is not a bug but is a consequence
    of how WordNet works. The word sense returned will always be the first
    word sense in a synset (synonym set) to which the original word belongs.

   Usage
    wsd.pl --context FILE --format FORMAT [--scheme SCHEME] [--type MEASURE]
    [--config FILE] [--stoplist FILE] [--window INT] [--contextScore NUM]
    [--pairScore NUM] [--outfile FILE] [--trace INT] [--forcepos]
    [--nocompoundify] | --help | --version

    The format option specifies one of the three different formats supported
    by wsd.pl. The three formats are:

    raw Raw text that is not part of speech tagged. This text should be
        formatted so that there is one sentence per line, one line per
        sentence. For example:


CHANGES
NAME
    CHANGES - Revision history for WordNet::SenseRelate::AllWords

DESCRIPTION
    0.18
        Date : April 30, 2009

        1)  Added following options to wsd.pl

            --usemono enable assigning the only available sense to monosemy
            words

            This is a change in AllWords.pm, where we assign the available
            sense to monosemy words. This results in increased precision and
            overall fmeasure because many measures can not get these words
            right as they can't find relatedness with the surrounding words.

        2) Added following options to allwords-scorer2.pl

            1. --word score only this specific instance
            2. --exceptword score all instances except for these instances
            3. --score poly score only polysemes instances
            4. --score s1nc score only the instances where the most frequent
            sense is not correct
            5. --score n score only the instances having n number of senses

            These options are added in order to carry different experiments.

        3)  Fixed the script calculate-corpus-stats.pl. This is a perl
            script that gives corpus statistics given a semcor-reformatted
            corpus file.

    0.17
        Date : April 2, 2009

        1)  Fixed AllWords.pm for window=2. If it is the first word in a
            sentence and the window size is 2, we now consider a word at the
            right of the target word. Otherwise it will not be assigned any
            sense. In the previous version we were simply doing sense1 for
            such scenario.

        2)  Added a script calculate-corpus-stats.pl to utils/ which gives
            statistics of the reformatted corpus.

    0.16
        Date : March 16, 2009

        1)  Fixed wsd.pl and the web interface output for multiword
            compounds ( for example, by_and_large).

        2)  Punctuation fix for the raw format. If the format is raw and if
            the context contains words i.e. or el_al., then the period is
            not removed.

        3)  If format is raw and the context contains user identified
            compound words (for example St._Petersburg), then the
            punctuation is not removed. Please refer README for details.

        4)  Mapped Penn Treebank IN tag to WordNet 'r'(adverb) part of
            speech tag because most of the words in this class have adverb
            sense in WordNet.

    0.15
        Date : February 14, 2009

        1)  Fixed the links of the default stoplist and nocompoundify option
            on the web interface.

    0.14
        Date : February 13, 2009

        1)  Added allwords-scorer2.pl script for scoring. This script is
            modeled after score2 C program
            (http://www.senseval.org/senseval3/scoring). scorer2 C program
            is no more used for scoring.

        2)  Changed wsd.pl and the web interface output format. The output
            now displays the original text. For example, if the input text
            is 'parking_ticket#n are#v expensive#a' then the output will be
            'parking_tickets#n#1 are#v#1 expensive#a#1'. Modified t/wsd.t
            for this change.

        3)  Removed config file option from the web interface

        4)  Added --nocompoundify option in wsd.pl and the web interface

        5)  Modified the default stoplist used by the web interface
            (web/cgi-bin/allwords/user_data/default-stoplist-raw.txt) and
            replicated it to samples/default-stoplist-raw.txt to be used by
            wsd.pl.

        6)  Fixed the ForcePos option on the webinterface.

        7)  Fixed the web interface crash when the user didn't provide any
            stoplist

        8)  Modified scorer2-format.pl to skip the entries where no sense is
            assigned.

        9)  --scorer2 option is no more provided in wsd-experiments.pl and
            the script calls utils/allwords-scorer2.pl.

    0.13
        Date : January 22, 2009

        1)  Setup for wsd.pl experiments. Added 2 scripts scorer2-sort.pl
            and wsd-experiments.pl. scorer2-sort.pl sorts the output of
            scorer2-format.pl so that it can be used as input to the
            Senseval scorer2 program. wsd-experiments.pl runs wsd
            experiments.

        2)  Documentation changes in wsd.pl, scorer2-format.pl,
            semcor-reformat.pl.

        3)  Makefile changes to install newly added perl scripts.

    0.12
        Date : November 03, 2008

        1)  Fixed client server model for the web interface. The allwords
            server can run on a different machine, from where your web
            server is running.

        2)  User data is now stored on the client machine so that it is
            easily accessible to the user.

        3)  Added a mandatory --logfile parameter alomg with other optional
            parameters to allwords_server.pl

        4)  Removed --outfile option from wsd.pl and added --glosses options
            to display glosses

        5)  Moved default-stoplist-raw.txt from
            WordNet-SenseRelate-AllWords/web/cgi-bin/allwords to
            /WordNet-SenseRelate-AllWords/web/cgi-bin/allwords/user_data

    0.11
        Date : June 17, 2008

        1)  Removed compounding logic from AllWords.pm. Compounding is now
            done using WordNet::Tools compoundify. The clients, wsd.pl and
            allwords_server.pl don't call compoundify any more. They creat
            WordNet::Tools object and pass it as a parameter while creating
            AllWords object.

        2)  Moved sentence splitter from wsd.pl to a util program,
            sentence_split.pl. All input formats will now assume that input
            is one sentence per line, one line per sentence.

        3)  Removed parsed format test case from t/wsd.t

        4)  Context was considered as a single line if more than one line

utils/wsd.pl
#!/usr/bin/perl
# $Id: wsd.pl,v 1.36 2009/04/30 22:20:26 kvarada Exp $

use strict;
use warnings;

use WordNet::SenseRelate::AllWords;
use WordNet::QueryData;
use WordNet::Tools;
use Getopt::Long;

our $measure = 'WordNet::Similarity::lesk';
our $mconfig;
our $contextf;
our $stoplist;
our $window = 3;
our $contextScore = 0;
our $pairScore = 0;
our $glosses;
our $nocompoundify;
our $usemono;
our $trace;
our $help;
our $version;
our $scheme = 'normal';
our $outfile;
our $forcepos;
our $val;
our $i=0;
our $j=0;


our $format; # raw|tagged|wntagged
my $OK_CHARS='-a-zA-Z0-9_\'\n ';

my $ok = GetOptions ('type|measure=s' => \$measure,
		     'config=s' => \$mconfig,
		     'context=s' => \$contextf,
		     'stoplist=s' => \$stoplist,
		     'window=i' => \$window,
		     'pairScore=f' => \$pairScore,
		     'contextScore=f' => \$contextScore,
		     'scheme=s' => \$scheme,
		     forcepos => \$forcepos,
		     glosses => \$glosses,
		     nocompoundify => \$nocompoundify,
		     usemono => \$usemono,
		     'trace=i' => \$trace,
		     help => \$help,
		     version => \$version,
		     'outfile=s' => \$outfile,
		     'format=s' => \$format,
		     );
$ok or exit 1;

if ($help) {
    showUsage ("Long");
    exit;
}

if ($version) {
    print "wsd.pl - assign a sense to all words in a context\n";
    print 'Last modified by : $Id: wsd.pl,v 1.36 2009/04/30 22:20:26 kvarada Exp $';
    print "\n";
    exit;
}

unless (defined $contextf) {
    showUsage ();
    exit 1;
}

unless ($format
        and (($format eq 'raw') or 
	     ($format eq 'tagged') or 
	     ($format eq 'wntagged'))) {
    showUsage ();
    exit 1;
}

unless ($scheme and (($scheme eq 'normal') or ($scheme eq 'random')
		     or ($scheme eq 'sense1') or ($scheme eq 'fixed'))) {
    showUsage ();
    exit 1;
}

#my $istagged = isTagged ($contextf);
my $istagged = $format eq 'tagged' ? 1 : 0;

if ($window < 2) {
    print STDERR "Error: the window must be 2 or larger!\n\n";
    exit 1;
}

    print STDERR "Current configuration:\n";
    print STDERR "    context file  : $contextf\n";
    print STDERR "    format        : $format\n";
    print STDERR "    scheme        : $scheme\n";
    print STDERR "    tagged text   : ", ($istagged ? "yes" : "no"), "\n";

    if (($scheme eq 'normal') or ($scheme eq 'fixed')) {
	# these items are only relevent to normal mode (not sense1 or random)
	print STDERR "    measure       : $measure\n";
	print STDERR "    window        : ", $window, "\n";
	print STDERR "    contextScore  : ", $contextScore, "\n";
	print STDERR "    pairScore     : ", $pairScore, "\n";
	print STDERR "    measure config: ", ($mconfig ? $mconfig : '(none)'), "\n";
	print STDERR "    glosses       : ", ($glosses ? "yes" : "no"), "\n";
	print STDERR "    nocompoundify : ", ($nocompoundify ? "yes" : "no"), "\n";
	print STDERR "    usemono      : ", ($usemono ? "yes" : "no"), "\n";
	print STDERR "    trace         : ", ($trace ? $trace : "no"), "\n";
	print STDERR "    forcepos      : ", ($forcepos ? "yes" : "no"), "\n";
    }

    print STDERR "    stoplist      : ", ($stoplist ? $stoplist : '(none)') , "\n";

local $| = 1;
print STDERR "Loading WordNet... ";
my $qd = WordNet::QueryData->new;
print STDERR "done.\n";

#...........................................................................
#
# Compoundifying is done using compoundify method of WordNet::Tools.
# The reason there was a --compounds option is that previoulsy we did
# not have a centralized compoundify module like we do now (WordNet::Tools)
# So each program would do their own compound identification.
# We think that this is no longer a good idea and hence we removed 
# --compounds option.
# WordNet::Tools object is passed while creating AllWords object. 
# AllWords.pm calls compoundify of using WordNet::Tools object reference.
# 
#...........................................................................

my $wntools = WordNet::Tools->new($qd);
$wntools or die "\nCouldn't construct WordNet::Tools object"; 

# options for the WordNet::SenseRelate constructor
my %options = (wordnet => $qd,
		wntools => $wntools,
	       measure => $measure,
	       );
$options{config} = $mconfig if defined $mconfig;
$options{stoplist} = $stoplist if defined $stoplist;
$options{trace} = $trace if defined $trace;
$options{pairScore} = $pairScore if defined $pairScore;
$options{contextScore} = $contextScore if defined $contextScore;
$options{outfile} = $outfile if defined $outfile;
$options{forcepos} = $forcepos if defined $forcepos;
$options{nocompoundify} = $nocompoundify if defined $nocompoundify;
$options{usemono} = $usemono if defined $usemono;
$options{wnformat} = 1 if $format eq 'wntagged';

my $sr = WordNet::SenseRelate::AllWords->new (%options);



open (FH, '<', $contextf) or die "Cannot open '$contextf': $!";

my @sentences;
if ($format eq 'raw') {
    local $/ = undef;
    my $input = <FH>;

#...........................................................................
#
# Removed splitSentences. We do not do any kind of sentence 
# splitting in wsd.pl. It is required the input to already be sentence 
# boundary processed. A small utility program utils/sentence_split.pl 
# is provided in case needed. 
# The text is cleaned in a way that is consistent with the Web interface
#
#...........................................................................
	$input =~ s/\r+//g;	
	@sentences = split(/\n+/,$input);
    undef $input;
    foreach my $sent (@sentences) {
	$sent = cleanLine ($sent);
	if ($sent !~ /[a-zA-Z0-9]/) {
		die "\nSorry. Your context should contain atleast one alphanumeric character\n";
	}
    }
}
else {
    @sentences = <FH>;
}

close FH;

foreach my $sentence (@sentences) {
    my @context = split /\s+/, $sentence;
    next unless scalar @context > 0;
    pop @context while !defined $context[$#context];
	
    my @res = $sr->disambiguate (window => $window,
				 tagged => $istagged,
				 scheme => $scheme,
				 context => [@context]);

	if($glosses)
	{
		print STDOUT join (' ', @context), "\n";
		print STDOUT join (' ', @res), "\n";
	}
	for($i=0,$j=0; $i<=$#res ; $i++,$j++)
	{
		   my $val;
  		   my $tagindex=index($res[$i],"#");
		   my $tag=substr $res[$i], $tagindex;
		   
		   if($format eq 'raw')
		   {
			 if($res[$i] =~ /\_/ && $context[$j] !~ /\_/)
			 {
				my $count = ($res[$i] =~ tr/\_//);
				$val=$res[$i];
				$j=$j+$count;
			 }else{
				$val=$context[$j].$tag;
			 }
		   }
		   elsif($format eq 'tagged')
		   {
			 my ($tw,$tt)= ( $context[$j] =~ /(\S+)\/(\S+)/);
			 $val=$tw.$tag;
		   }	
  		   elsif($format eq 'wntagged')
		   {
			 my ($tw,$tt)= split /\#/, $context[$j];
			 $val=$tw.$tag;
		   }	

		   if($glosses)
		   {
			if($val =~ /\#o/ )
			{
				print STDOUT "\n$val : stopword";
			}
			elsif($val =~ /\#ND/) 
			{
				print STDOUT "\n$val : not in WordNet\n";
			}
			elsif($val =~ /\#NR/)
			{
				print STDOUT "\n$val: No relatedness found with the surrounding words\n";
			}
			elsif($val =~ /\#IT/)
			{
				print STDOUT "\n$val: Invalid Tag\n";
			}
			elsif($val =~ /\#NT/)
			{
				print STDOUT "\n$val: No Tag\n";
			}
			elsif($val =~ /\#CL/)
			{
				print STDOUT "\n$val: Closed Class Word\n";
			}
			elsif($val =~ /\#MW/)
			{
				print STDOUT "\n$val : Missing Word\n";
			}
			else
			{
				my ($gloss) = $qd->querySense ($res[$i], "glos");
				print STDOUT "\n$val : $gloss\n";
			}
		   }
		   else
		   {
			print $val." ";
		   }	   	
		}
 	       print "\n";

		if ($trace) 
		{
			my $tstr = $sr->getTrace ();
			print $tstr, "\n";
    		}
}

sub isTagged
{
    my $file = shift;
    open FH, '<', $file or die "Cannot open context file '$file': $!";
    my @words;
    while (my $line = <FH>) {
	chomp $line;
	push @words, split (/\s+/, $line);
	last if $#words > 20;
    }
    close FH;

    my $tag_count = 0;
    foreach my $word (@words) {
	$tag_count++ if $word =~ m|/\S|;
    }
    my $ratio = $tag_count / scalar @words;

    # we consider the corpus to be tagged if we found that 70% or more
    # of the first 20 words were tagged (70% is somewhat of an arbitrary
    # value).
    return 1 if $ratio > 0.7;
    return 0;
}

sub cleanLine
{
    	my $line = shift;
	chomp($line);
	my @words=split(/ +/,$line);
	foreach my $word (@words){
		next if($word eq "i.e." || $word eq "ie." || $word eq "et_al." || $word eq "al.");
		$word =~ s/([A-Z])/\L$1/g;
		if ($word =~ m/_/){
			$word =~ s/[.|!|?|,|;]+$/ /;
		}
		else{
			$word =~ s/[^$OK_CHARS]/ /g;
		}
	}
	return join (' ', @words);
}

sub showUsage
{
    my $long = shift;
    print "Usage: wsd.pl --context FILE --format FORMAT [--scheme SCHEME]\n";
    print "              [--type MEASURE] [--config FILE] \n";
    print "              [--stoplist file] [--window INT] [--contextScore NUM]\n";
    print "              [--pairScore NUM] [--outfile FILE] [--trace INT] \n";
    print "              [--glosses][--forcepos][--nocompoundify][--usemono] \n";
    print "              | {--help | --version}\n";

    if ($long) {
	print "Options:\n";
	print "\t--context FILE       a file containing the text to be disambiguated\n";
	print "\t--format FORMAT      type of --context ('raw', 'tagged',\n";
       print "\t                     or 'wntagged')\n";
	print "\t--scheme SCHEME      disambiguation scheme to use. ('normal', \n";
	print "\t                     'fixed', 'sense1', or 'random')\n";
	print "\t--type MEASURE       the relatedness measure to use\n";
	print "\t--config FILE        a configuration file for the relatedness measure\n";
	print "\t--stoplist FILE      a file of regular expressions that define\n";
	print "\t                     the words to be excluded from --context\n";
	print "\t--window INT         window of context will include INT words\n";
	print "\t                     in all, including the target word.\n";
	print "\t--contextScore NUM   the  minimum required of a winning score\n";
	print "\t                     to assign a sense to a target word\n";
	print "\t--pairScore NUM      the minimum pairwise threshold when\n";
	print "\t                     measuring target and word in window\n";
	print "\t--outfile FILE       create a file with one word-sense per line\n";
	print "\t--trace INT          set trace levels. greater values show more\n";
	print "\t                     detail. may be summed to combine output. \n";
	print "\t--glosses            show glosses of each disambiguated word\n";
       print "\t--forcepos           force all words in window of context\n";
       print "\t                     to be same pos as target (pos coercion)\n";
	print "\t                     are assigned\n";
       print "\t--nocompoundify      disable compoundify\n";
       print "\t--usemono            enable assigning the only available sense to monosemy words\n";
	print "\t--help               show this help message\n";
	print "\t--version            show version information\n";
    }
}

__END__

=head1 NAME

wsd.pl - automatically assign a meaning to every word in a text

=head1 SYNOPSIS

 wsd.pl --context FILE --format FORMAT [--scheme SCHEME] [--type MEASURE] 
           [--config FILE] [--stoplist FILE] 
           [--window INT] [--contextScore NUM] [--pairScore NUM] 
           [--outfile FILE] [--trace INT] [--forcepos] [--nocompoundify] [--usemono]
		| --help | --version

=head1 DESCRIPTION

Disambiguates each word in the context file using the specified relatedness
measure (or WordNet::Similarity::lesk if none is specified).

=head1 OPTIONS

N.B., the I<=> sign between the option name and the option parameter is
optional.

=over

=item --context=B<FILE>

The input file containing the text to be disambiguated.  This
"option" is required.

=item --format=B<FORMAT>

The format of the input file. For all formats there must be one sentence
per line, one line per sentence.  Valid values are:

=over

=item raw

The input is raw text. Compounds will be identified, punctuation is 
ignored. 

=item tagged 

The input has been part-of-speech tagged with Penn Treebank tags.
Compounds are not identified, and untagged words are ignored. 

=item wntagged

The input has been part-of-speech tagged with WordNet tags (n, v, a, r).
Compounds are not identified, and untagged words are ignored. 

=back

=item --scheme=B<SCHEME>

The disambiguation scheme to use.  Valid values are "normal", "fixed",
"sense1", and "random". The default is "normal".  In fixed mode, once a word
is assigned a sense number, other senses of that word won't be considered
when disambiguating words to the right of that context word.  For example,
if the context is

  dogs run very fast

and 'dogs' has been assigned sense number 1, only sense 1 of dogs will
be used in computing relatedness values when disambiguating 'run', 'very',
and 'fast'.

WordNet sense 1
disambiguation  guesses that the correct sense for each word is the
first sense in WordNet because the senses of words in WordNet are
ranked according to frequency.   
The first sense is more likely than the second, the second is more likely  
than the third, etc. Random selects one of the possible senses of the 
target word randomly. 

=item --type=B<MEAURE>

The relatedness measure to be used.  The default is WordNet::Similarity::lesk.

=item --config=B<FILE>

The name of a configuration file for the specified relatedness measure.

=item --stoplist=B<FILE>

A file containing regular expressions (as understood by Perl), surrounded by
by slashes (e.g. /\d+/ removes any word containing a digit [0-9]).  Any word
in the text to be disambiguated that matches one of the regular  
expressions in the file is removed.  Each regular expression must be on  
its own line, and any trailing whitespace is ignored.

Care must be taken when crafting a stoplist.  For example, it is tempting
to use /a/ to remove the word 'a', but that expression would result in
all words containing the lowercase letter a to be removed.  A better
alternative would be /\ba\b/.

=item --window=B<INTEGER>

Defines the size of the window of context.  The default is 4.  A window
size of N means that there will be a total of N words in the context
window, including the target word.  If N is a (positive) even number,
then there will be one more word on the left side of the target word
than on the right.

For example, if the window size is 4, then there will be two words on
the left side of the target word and one on the right.  If the window
is 5, then there will be two words on each side of the target word.

The minimum window size is 2.  A smaller window would mean that there
were no context words in the window.

=item --contextScore=B<REAL>

If no sense of the target word achieves this minimum score, then
no winner will be projected (e.g., it is assumed that there is
no best sense or that none of the senses are sufficiently related
to the surrounding context).  The default is zero.

=item --pairScore=B<REAL>

The minimum pairwise score between a sense of the target word and
the best sense of a context word that will be used in computing
the overall score for that sense of the target word.  Setting this
to be greater than zero (but not too large) will reduce noise.
The default is zero.

=item --outfile=B<FILE>

The name of a file to which output should be sent. This file will display 
one word and its sense per line.  

=item --trace=B<INT>

Turn tracing on/off.  A value of zero turns tracing off, a non-zero value
turns tracing on.  The different trace levels can be added together
to see the combined traces.  The trace levels are:

  1 Show the context window for each pass through the algorithm.

  2 Display winning score for each pass (i.e., for each target word).

  4 Display the non-zero scores for each sense of each target
    word (overrides 2).

  8 Display the non-zero values from the semantic relatedness measures.

 16 Show the zero values as well when combined with either 4 or 8.
    When not used with 4 or 8, this has no effect.

 32 Display traces from the semantic relatedness module.

=item --forcepos

Turn part of speech coercion on.  POS coercion attempts to force other words
in the context window to be of the same part of speech as the target word.
If the text is POS tagged, the POS tags will be ignored.
POS coercion  may be useful when using a measure of semantic similarity that
only works with noun-noun and verb-verb pairs.

=item --nocompoundify

Disable compoundifying. By default AllWords.pm compoundifes the input raw text. 
Using this option will disable this. 

=item --usemono

If this flag is on the only available sense is assignsed to the usemono words. 
By default this flag is off. 

=back

=head1 SEE ALSO

 L<WordNet::SenseRelate::AllWords>

The main web page for SenseRelate is

 L<http://senserelate.sourceforge.net/>

There are several mailing lists for SenseRelate:

 L<http://lists.sourceforge.net/lists/listinfo/senserelate-users/>

 L<http://lists.sourceforge.net/lists/listinfo/senserelate-news/>

 L<http://lists.sourceforge.net/lists/listinfo/senserelate-developers/>

=head1 AUTHORS

 Jason Michelizzi 

 Ted Pedersen, University of Minnesota, Duluth
 E<lt>tpederse at d.umn.eduE<gt>

=head1 BUGS

Please report to senserelate-users mailing list. 

=head1 COPYRIGHT

Copyright (C) 2004-2008 Jason Michelizzi and Ted Pedersen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

=cut
